TTS/tts/models/naturalspeech2.py

```diff
@@ -0,0 +1,1276 @@
+            "token_ids": token_ids,
+            "token_len": len(token_ids),
+            "wav": wav,
+            "wav_file": wav_filename,
+            "language_name": item["language"],
+            "audio_unique_name": item["audio_unique_name"],
+            "pitch": pitch,
+        }
+
+    @property
+    def lengths(self):
+        lens = []
+        for item in self.samples:
+            _, wav_file, *_ = _parse_sample(item)
+            audio_len = os.path.getsize(wav_file) / 16 * 8  # assuming 16bit audio
+            lens.append(audio_len)
+        return lens
+
+    def collate_fn(self, batch):
+        """
+        Return Shapes:
+            - tokens: :math:`[B, T]`
+            - token_lens :math:`[B]`
+            - token_rel_lens :math:`[B]`
+            - waveform: :math:`[B, 1, T]`
+            - waveform_lens: :math:`[B]`
+            - waveform_rel_lens: :math:`[B]`
+            - language_names: :math:`[B]`
+            - audiofile_paths: :math:`[B]`
+            - raw_texts: :math:`[B]`
+            - audio_unique_names: :math:`[B]`
+        """
+        # convert list of dicts to dict of lists
+        B = len(batch)
+        batch = {k: [dic[k] for dic in batch] for k in batch[0]}
+
+        _, ids_sorted_decreasing = torch.sort(
+            torch.LongTensor([x.size(1) for x in batch["wav"]]), dim=0, descending=True
+        )
+
+        max_text_len = max([len(x) for x in batch["token_ids"]])
+        token_lens = torch.LongTensor(batch["token_len"])
+        token_rel_lens = token_lens / token_lens.max()
+
+        wav_lens = [w.shape[1] for w in batch["wav"]]
+        wav_lens = torch.LongTensor(wav_lens)
+        wav_lens_max = torch.max(wav_lens)
+        wav_rel_lens = wav_lens / wav_lens_max
+
+        # format F0
+        pitch = prepare_data(batch["pitch"])
+        pitch = torch.FloatTensor(pitch)[:, None, :].contiguous()  # B x 1 xT
+
+        # latents_lens = [w.shape[2] for w in batch["latents"]]
+        # latents_lens = torch.LongTensor(latents_lens)
+        # latents_lens_max = torch.max(latents_lens)
+        # latents_rel_lens = latents_lens / latents_lens_max
+
+        token_padded = torch.LongTensor(B, max_text_len)
+        wav_padded = torch.FloatTensor(B, 1, wav_lens_max)
+        token_padded = token_padded.zero_() + self.pad_id
+        wav_padded = wav_padded.zero_() + self.pad_id
+        for i in range(len(ids_sorted_decreasing)):
+            token_ids = batch["token_ids"][i]
+            token_padded[i, : batch["token_len"][i]] = torch.LongTensor(token_ids)
+
+            wav = batch["wav"][i]
+            wav_padded[i, :, : wav.size(1)] = torch.FloatTensor(wav)
+
+            # latents_emb = batch["latents"][i]
+            # latents_padded[i, :, : latents_emb.size(2)] = torch.FloatTensor(latents_emb)
+        # Extract discrete codes from EnCodec
+        emb, codes, _ = self.encodec(wav_padded, return_encoded=True)
+        emb = emb.squeeze(1)
+        emb = emb.transpose(1, 2)
```

Comment:
If you do this in `format_batch_on_device` you can speed up the data loading significantly.